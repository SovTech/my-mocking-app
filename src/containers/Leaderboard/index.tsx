import * as React from 'react';
import { Query } from 'react-apollo';
import { ALL_USERS } from '../../graphql/queries';
import styled from 'styled-components';

interface User {
    email: string;
    firstName: string;
}

const UserRow = styled.div`
  color: white;
  flex-direction: row;
  display: flex;
  justify-content: space-between;
  width: 300px;
  max-width: 100%;
  
  span {
      display: flex;
      flex: 1;
  }
`;

export default class Leaderboard extends React.Component {
    render() {
        return (
            <div>
                <Query query={ALL_USERS}>
                    {({loading, error, data}) => {
                        if (loading) {
                            return <h2>Loading...</h2>;
                        }
                        if (error) {
                            return <h2>Error 😭</h2>;
                        }
                        if (!data.allUsers) {
                            return <h2>There are no users to display 😭</h2>;
                        }

                        return data.allUsers.map((user: User, index: number) =>
                            <UserRow key={user.email}>
                                <span>{`${index + 1}.`}</span>
                                <span>
                                    {user.firstName.toUpperCase()
                                    + (index === 0 ? '🏆' : '')
                                    + (index === data.allUsers.length - 1 ? '🥔👑' : '')}
                                </span>
                            </UserRow>
                        );
                    }}
                </Query>
            </div>
        );
    }
}
