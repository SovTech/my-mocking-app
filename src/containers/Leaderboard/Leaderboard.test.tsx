import * as React from 'react';
import Leaderboard from './index';
import { ApolloProvider } from 'react-apollo';
import { ALL_USERS } from '../../graphql/queries';
import * as renderer from 'react-test-renderer';
import { mock } from '../../testing/graphql';

const App = () => (
    <ApolloProvider client={mock.client as any}>
        <Leaderboard/>
    </ApolloProvider>
);

describe('Leaderboard', () => {
    beforeEach(() => mock.reset());
    it('renders the leaderboard', () => {
        mock.expect(ALL_USERS).reply({
            allUsers: [
                {id: '1', firstName: 'Jonathan', email: 'jono@example.com'},
                {id: '2', firstName: 'Deepak', email: 'deepak@example.com'}
            ]
        });

        const tree = renderer.create(App()).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders loading state', () => {
        mock.expect(ALL_USERS).loading(true);

        const tree = renderer.create(App()).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders errors when API fails', () => {
        mock.expect(ALL_USERS).fail('everything is terrible');

        const tree = renderer.create(App()).toJSON();
        expect(tree).toMatchSnapshot();
    });
});