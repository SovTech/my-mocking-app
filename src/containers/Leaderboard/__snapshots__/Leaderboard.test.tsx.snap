// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Leaderboard renders errors when API fails 1`] = `
<div
  className="sc-cSHVUG dTqyNa"
>
  <div
    className="sc-kAzzGY jBsOLT"
  >
    <h2
      className="sc-bxivhb fvRKrT"
    >
      Error 😭
    </h2>
  </div>
  <div
    className="sc-chPdSV esFuas"
  >
    v
    x.x.x
  </div>
</div>
`;

exports[`Leaderboard renders loading state 1`] = `
<div
  className="sc-cSHVUG dTqyNa"
>
  <div
    className="sc-kAzzGY jBsOLT"
  >
    <h2
      className="sc-bxivhb fvRKrT"
    >
      Loading...
    </h2>
  </div>
  <div
    className="sc-chPdSV esFuas"
  >
    v
    x.x.x
  </div>
</div>
`;

exports[`Leaderboard renders the leaderboard 1`] = `
<div
  className="sc-cSHVUG dTqyNa"
>
  <div
    className="sc-kAzzGY jBsOLT"
  >
    <div
      className="sc-dxgOiQ bOTcL"
    >
      <span>
        1.
      </span>
      <span>
        JONATHAN🏆
      </span>
    </div>
    <div
      className="sc-dxgOiQ bOTcL"
    >
      <span>
        2.
      </span>
      <span>
        DEEPAK🥔👑
      </span>
    </div>
  </div>
  <div
    className="sc-chPdSV esFuas"
  >
    v
    x.x.x
  </div>
</div>
`;
